src/library/dao/macrodao.cpp

```diff
@@ -0,0 +1,111 @@
+#include "library/dao/macrodao.h"
+
+#include <QDebug>
+#include <QSqlQuery>
+
+#include "library/queryutil.h"
+
+bool MacroDAO::saveMacro(TrackId trackId, Macro* macro, int slot) const {
+    QSqlQuery query(m_database);
+    if (macro->getId() == -1) {
+        query.prepare(QStringLiteral(
+                "INSERT INTO macros "
+                "(track_id, slot, label, state, content) "
+                "VALUES "
+                "(:track_id, :slot, :label, :state, :content)"));
+        if (slot == 0) {
+            slot = getFreeSlot(trackId);
+        }
+    } else {
+        VERIFY_OR_DEBUG_ASSERT(slot != 0) {
+            slot = getFreeSlot(trackId);
+        }
+        query.prepare(QStringLiteral(
+                "UPDATE macros SET "
+                "track_id=:track_id,"
+                "slot=:slot,"
+                "label=:label,"
+                "state=:state,"
+                "content=:content"
+                " WHERE id=:id"));
+        query.bindValue(":id", macro->getId());
+    }
+    query.bindValue(":track_id", trackId.toVariant());
+    query.bindValue(":slot", slot);
+    query.bindValue(":label", macro->getLabel());
+    query.bindValue(":state", macro->getState().operator int());
+    query.bindValue(":content", Macro::serialize(macro->getActions()));
+
+    if (query.exec()) {
+        if (macro->getId() == -1) {
+            macro->setId(query.lastInsertId().toInt());
+        }
+        macro->setDirty(false);
+        return true;
+    } else {
+        LOG_FAILED_QUERY(query);
+        return false;
+    }
+}
+
+void MacroDAO::saveMacros(TrackId trackId, QMap<int, MacroPtr> macros) const {
+    for (auto e : macros.toStdMap()) {
```

Comment:
What is `e`? Why to you create a standard map first?

The standard Qt way to iterate over the map is:

```cpp
auto it = map.constBegin();
while (it != map.constEnd()) {
    //qDebug() << it.key() << ": " << it.value();
    ++it;
}
```

That doesn't create a copy and is more readable than `.first`/`.second` IMHO.

Reply:
I think I copied that somewhere, I'll adjust it, didn't know about that :)